com.unity.render-pipelines.universal/Runtime/UniversalRenderPipeline.cs

```diff
@@ -22,14 +23,181 @@ public LightweightRenderPipeline(LightweightRenderPipelineAsset asset)
+            return new ProfilingScope(null, scopeSampler);
+        }
+
+        // Convenience overload for switching between cached and static sampler
+        public static ProfilingScope GetGPUScope(CommandBuffer cmd, ProfilingSampler scopeSampler)
+        {
+            return new ProfilingScope(cmd, scopeSampler);
+        }
+
+        // Get a cached ProfilingSampler or add into the cache if not found.
+        // NOTE:
+        // Caching has a tiny overhead (hashing).
+        // Typically no allocations, but might allocate once on a cache miss for Add operation.
+        public static ProfilingSampler TryGetOrAddSampler(string samplerName)
+        {
+            // TODO: do profiling levels
+            //#define UNIVERSAL_PROFILING_MINIMAL
+            #if UNIVERSAL_PROFILING_MINIMAL
+                // Disable caching and return a static sampler for all scopes.
+                return m_UnknownSampler;
+            #else
+                ProfilingSampler ps = null;
+                bool exists = m_StringSamplerCache.TryGetValue(samplerName, out ps);
+                if (!exists)
+                {
+                    ps = new ProfilingSampler(samplerName);
+                    m_StringSamplerCache.Add(samplerName, ps);
+                }
+
+                return ps;
+            #endif
+        }
+
+        // Get a cached ProfilingSampler, or add into the cache if not found, using a hash value for any object.
+        // Adds a sampler with a name of form `categoryName + dynamicName` using a generic key value.
+        // NOTE:
+        // Caching has some overhead (hashing (twice)).
+        // Typically no allocations, but might allocate once on a cache miss for Add operation.
+        //
+        // <param name"hashKey"> Generic integer used as the cache key. Typically a hashed value.
+        // <param name"categoryName"> Shared part of the dynamic name in "CategoryName" + "DynamicName".
+        // <param name"dynamicName"> Unique detail part of the dynamic name in "CategoryName" + "Dynamicname".
+        public static ProfilingSampler TryGetOrAddSampler(int hashKey, string categoryName, string detailName)
+        {
+            #if UNIVERSAL_PROFILING_MINIMAL
+                // Disable caching and return a static sampler for all scopes.
+                return m_UnknownSampler;
+            #else
+                ProfilingSampler ps = null;
+                bool exists = m_IntSamplerCache.TryGetValue(hashKey, out ps);
+                if (!exists)
+                {
+                    ps = new ProfilingSampler( categoryName + detailName);
+                    m_IntSamplerCache.Add(hashKey, ps);
+                }
+
+                return ps;
+            #endif
+        }
+
+        // Specialization for camera loop to avoid allocations.
+        public static ProfilingSampler TryGetOrAddCameraSampler(Camera camera)
+        {
+            ProfilingSampler ps = null;
+            int cameraId = camera.GetHashCode();
```

Comment:
Maybe, this is actually bad/slow. How big is the camera class? What does GetHashCode() actually hash.

Reply:
Measured to be only single microseconds.